components/compositing/compositor.rs

```diff
@@ -705,25 +688,46 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
                 script_traits::ScriptToCompositorMsg::SendInitialTransaction(pipeline),
             ) => {
                 let mut txn = Transaction::new();
-                txn.set_display_list(
-                    WebRenderEpoch(0),
-                    None,
-                    Default::default(),
-                    (pipeline, Default::default()),
-                    false,
-                );
-
-                self.generate_frame(&mut txn);
+                txn.set_display_list(WebRenderEpoch(0), (pipeline, Default::default()));
+                self.generate_frame(&mut txn, RenderReasons::SCENE);
                 self.webrender_api
                     .send_transaction(self.webrender_document, txn);
             },
 
             ForwardedToCompositorMsg::Layout(
-                script_traits::ScriptToCompositorMsg::SendScrollNode(point, scroll_id),
+                script_traits::ScriptToCompositorMsg::SendScrollNode(
+                    pipeline_id,
+                    point,
+                    external_scroll_id,
+                ),
             ) => {
+                let pipeline_id = PipelineId::from_webrender(pipeline_id);
+                let pipeline_details = match self.pipeline_details.get_mut(&pipeline_id) {
+                    Some(details) => details,
+                    None => return,
+                };
+
+                let offset = LayoutVector2D::new(point.x, point.y);
+                if !pipeline_details
+                    .scroll_tree
+                    .set_scroll_offsets_for_node_with_external_scroll_id(
+                        external_scroll_id,
+                        -offset,
+                    )
+                {
+                    warn!("Could not scroll not with id: {external_scroll_id:?}");
+                    return;
+                }
+
                 let mut txn = Transaction::new();
-                txn.scroll_node_with_id(point, scroll_id, ScrollClamping::NoClamping);
-                self.generate_frame(&mut txn);
+                txn.set_scroll_offsets(
+                    external_scroll_id,
+                    vec![SampledScrollOffset {
+                        offset,
```

Comment:
The scroll tree and WebRender have opposite interpretations of the scroll offset unfortunately. I'd like to fix that in a followup, because it is pretty confusing indeed.